### PR TITLE
Fix constraint interval derivation for commutative operators

### PR DESCRIPTION
**Pull request by dgkimura**
ORCA constraint property is used to detect constraint contradiction in
plan subtrees. If a contradiction is detected then the optimizer prunes
the subtree. In constraint property deriviation the predicate is used to
construct a constraint internval.

Issue is that commit https://github.com/greenplum-db/gpdb/commit/6df178bec195abc0c7aa7858c382571a48153a74 allowed ORCA to consider compare operators
of the form (CONST op IDENT) when constructing the constraint interval.
However, the underlying mechanism continued to make assumptions about
the order of the arguments. The problem with the aforementioned commit
is that it assumed the comparison operator types were same. Instead it
should have chosen to use the commutative comparison operator type.

Following example demonstrated this issue:

  ```sql
  CREATE TABLE t(sample_time timestamp);
  INSERT INTO t values ('2023-03-02 00:00:00');
  SELECT * FROM t
  WHERE '2023-03-02 00:00:00'::timestamp(0) + interval '1 month' >= sample_time AND
        '2023-03-02 00:00:00'::timestamp(0) - interval '1 month' <= sample_time;
  ```

ORCA preprocessor step PexprPruneEmptySubtrees checks the constraint
intervals and removes contradicting constraints. In this case, while
computing following portion of the constraint:
```
  +--CScalarCmp (>=)
     |--CScalarConst (733708800000000.000)
     +--CScalarIdent "sample_time" (0)
```

It created an incorrect contstraint interval:
```
  {"sample_time" (0), ranges: [733708800000000.000, inf) }
```

It should have been:
```
  {"sample_time" (0), ranges: (-inf, 733708800000000.000] }
```

Notice that the constraint form is (CONST >= IDENT). By using the
original comparison type the constructed interval range is incorrectly
inverted into (IDENT >= CONST). This is because the mechanism continues
to assume (IDENT op CONST). In this case, when both the ranges are
reversed they end up contradicting each other. In that case ORCA
believes there should be no results and produces corresponding empty
subtree.

### CODE REVIEW

File: src/backend/gporca/libgpopt/src/operators/CPredicateUtils.cpp [hunk 1]
```diff
@@ -991,19 +991,36 @@ CPredicateUtils::PexprINDFConjunction(CMemoryPool *mp,
 }
 
 // is the given expression a comparison between a scalar ident and a constant
+//
+// Expression may is of the forms:
+//   (IDENT op CONST)
+//   (CONST op' IDENT)
+//
+// If it is (COST op IDENT) then also check that the comparison operator is the
```
Comment (DevChattopadhyay) [inline src/backend/gporca/libgpopt/src/operators/CPredicateUtils.cpp:999]: COST->CONST

File: src/backend/gporca/libgpopt/src/operators/CPredicateUtils.cpp [hunk 2]
```diff
@@ -1034,6 +1052,14 @@ CPredicateUtils::FCompareIdentToConst(CExpression *pexpr,
 	{
 		pexprIdent = pexprRight;
 		pexprConst = pexprLeft;
+
+		CMDAccessor *md_accessor = COptCtxt::PoctxtFromTLS()->Pmda();
+		const IMDScalarOp *opSc = md_accessor->RetrieveScOp(
+			CScalarCmp::PopConvert(pexpr->Pop())->MdIdOp());
+		const IMDScalarOp *opScComm =
+			md_accessor->RetrieveScOp(opSc->GetCommuteOpMdid());
```
Comment (DevChattopadhyay) [inline src/backend/gporca/libgpopt/src/operators/CPredicateUtils.cpp:1060]: Just for knowledge - There might be a case where there is an operator without a commutative operator defined.What will happen in that case.

**Review by DevChattopadhyay — APPROVED** [review]
This looks good to me.

File: src/backend/gporca/libgpopt/src/operators/CPredicateUtils.cpp [hunk 3]
```diff
@@ -991,19 +991,36 @@ CPredicateUtils::PexprINDFConjunction(CMemoryPool *mp,
...
 BOOL
 CPredicateUtils::FCompareIdentToConst(CExpression *pexpr)
 {
 	CExpression *pexprIdent;
 	CExpression *pexprConst;
-
-	return FCompareIdentToConst(pexpr, pexprIdent, pexprConst);
+	IMDType::ECmpType cmptype;
+	return FCompareIdentToConst(pexpr, pexprIdent, pexprConst, cmptype) &&
+		   cmptype == CScalarCmp::PopConvert(pexpr->Pop())->ParseCmpType();
```
[the remark this answers is not in the document]
Comment (dgkimura) [reply src/backend/gporca/libgpopt/src/operators/CPredicateUtils.cpp:1009]: Yes, that was the expectation.  Felt this preserved old behavior enough while allowing wiggle room for when commutative operator is same regardless of order of operands.

[the remark this answers is not in the document]
Comment (THANATOSLAVA) [reply src/backend/gporca/libgpopt/src/operators/CPredicateUtils.cpp:1009]: Given `FCompareIdentToConst` is only used in `PciIntervalFromScalarCmp`, I wonder if we could return true on `5 > a` as well, and derive the interval from `a EcmptL 5`.

[the remark this answers is not in the document]
Comment (dgkimura) [reply src/backend/gporca/libgpopt/src/operators/CPredicateUtils.cpp:1009]: Yeah, I had thought about that.  I am open to looking at trying in 7X.  But for 6X I am more hesitant and I may actually prefer to bring back old behavior.

**Review by gpopt — COMMENTED** [review]
Instead, can we move the between range derivation after preprocess step (26), which reorders the scalar cmp operator?

**Comment by dgkimura** [comment]
On Monday, March 20, 2023 9:41 PM gpopt wrote:
> Instead, can we move the between range derivation after preprocess step (26), which reorders the scalar cmp operator?

Yeah, that works.  And it's simpler.  One 'meh' is that it adds another non-obvious dependency between preprocessor steps.  But maybe that's to be expected.  I like the idea though.  I'll pull on this string and see where it leads.  Thanks!

**Comment by dgkimura** [comment]
Alternative approach that reorders preprocessing steps: https://github.com/greenplum-db/gpdb/pull/15242.  It includes revert of previous support added for `CONST op IDENT` comparison expression in constraint interval construction.  Both approaches are different ways to solve the same issue.  Thanks @gpopt for the suggestion!

**Comment by dgkimura** [comment]
Closed in favor of alternative approach.  Thanks for reviews!

**Pull request closed without merging**